src/Mongolid/DataMapper/DataMapper.php

```diff
@@ -0,0 +1,280 @@
+     *
+     * @return \Mongolid\Cursor\Cursor
+     */
+    public function where($query = []): Cursor
+    {
+        $cursor = new Cursor(
+            $this->schema->entityClass,
+            $this->getCollection(),
+            'find',
+            [$this->prepareValueQuery($query)]
+        );
+
+        return $cursor;
+    }
+
+    /**
+     * Retrieve a database cursor that will return all documents as
+     * $this->schema->entityClass objects upon iteration
+     *
+     * @return \Mongolid\Cursor\Cursor
+     */
+    public function all(): Cursor
+    {
+        return $this->where([]);
+    }
+
+    /**
+     * Retrieve one $this->schema->entityClass objects that matches the given
+     * query
+     *
+     * @param  array $query MongoDB query to retrieve the document
+     *
+     * @return mixed First document matching query as an $this->schema->entityClass object
+     */
+    public function first($query)
+    {
+        $document = $this->getCollection()->findOne(
+            $this->prepareValueQuery($query)
+        );
+
+        if (! $document) {
+            return null;
+        }
+
+        $model = new $this->schema->entityClass;
+
+        foreach ($document as $key => $value) {
+            $model->$key = $value;
+        }
+
+        return $model;
+    }
+
+    /**
+     * Parses an object with SchemaMapper and the given Schema
+     *
+     * @param  mixed  $object
+     *
+     * @return array  Document
+     */
+    protected function parseToDocument($object)
+    {
+        $schemaMapper     = $this->getSchemaMapper();
+        $objectAttributes = $this->parseToArray($object);
+
+        $parsedDocument = $schemaMapper->map($objectAttributes);
+
+        if (isset($parsedDocument['_id']) && is_object($object)) {
+            $object->_id = $parsedDocument['_id'];
+        }
+
+        return $parsedDocument;
+    }
+
+    /**
+     * Returns a SchemaMapper with the $schema or $schemaClass instance
+     *
+     * @return SchemaMapper
+     */
+    protected function getSchemaMapper()
+    {
+        if (! $this->schema) {
+            $this->schema = Ioc::make($this->schemaClass);
+        }
+
+        return Ioc::make('Mongolid\DataMapper\SchemaMapper', [$this->schema]);
+    }
+
+    /**
+     * Parses an object to an array before sending it to the SchemaMapper
+     *
+     * @param  mixed $object
+     *
+     * @return array
+     */
+    protected function parseToArray($object): array
+    {
+        if (! is_array($object)) {
+            if (method_exists($object, 'getAttributes')) {
+                return $object->getAttributes();
+            }
+
+            return get_object_vars($object);
+        }
+
+        return $object;
+    }
+
+    /**
+     * Retrieves the Collection object.
+     *
+     * @return Collection
+     */
+    protected function getCollection(): Collection
+    {
+        $conn       = $this->connPool->getConnection();
```

Comment:
Be careful if `getConnection()` failed, it's can return null and you try to call method.